File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function(heights) {
    const n = heights.length;
    let stack = [-1];
    let maxArea = 0;
    for(let i=0; i<n; i++) {
        while(stack[stack.length-1] !== -1 && heights[stack[stack.length-1]] >= heights[i]) {
            const currentHeight = heights[stack.pop()];
            const currentWidth = i - stack[stack.length-1] -1;
            maxArea = Math.max(maxArea, currentHeight*currentWidth);
        }
        stack.push(i);
    }
    while(stack[stack.length-1] !== -1) {
        const currentHeight = heights[stack.pop()];
        const currentWidth = n - stack[stack.length-1] -1;
        maxArea = Math.max(maxArea, currentHeight*currentWidth);
    }
    return maxArea;
};

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        max_area = 0
        stack = []

        for i, h in enumerate(heights):
            start = i

            # FIX: Added to access the height element of the tuple
            while stack and stack[-1][1] > h:
                index, height = stack.pop()

                max_area = max(max_area, height * (i - index))

                start = index

            stack.append((start, h))

        for i, h in stack:
            max_area = max(max_area, h * (len(heights) - i))

        return max_area